daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Profile.scala

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import java.lang.System
+import java.util.ArrayList
+
+/** Class for profiling information collected by Speedy. We use [[AnyRef]] for
+  * the labels to avoid any conversions into a common label format at runtime
+  * since this might skew the profile. Instead, we convert the labels to strings
+  * when we write out the profile.
+  */
```

Comment:
Could you at least use an unboxed newtype or `scalaz.Tag` for the `AnyRef` here? This would not entail any allocations. In the former case (bespoke newtype) we could also use it to proscribe the set of types permitted to be used for labels to those supported by `Event#label`, again without implying any allocation. To carry the latter to its conclusion,

```scala
type Label = LabelModule.Module.T

sealed abstract class LabelModule {
  type T <: AnyRef
  private def apply(t: AnyRef): T
}
object LabelModule {
  val Module: LabelModule = new LabelModule {
    type T = AnyRef
    override private def apply(t: T) = t
  }

  // assumes -Xsource:2.13 in clients
  implicit def fromAllowed[T: Allowed](t: T): Label = Module(t)

  @implicitNotFound(msg = "Values of type ${T} are not known, renderable Labels")
  final class Allowed[-T] private()
  object Allowed {
    private[this] val allowAll = new Allowed[Any]
    implicit val anonClosure: Allowed[AnonymousClosure.type] = allowAll
    implicit val str: Allowed[String] = allowAll
    // and so on for other permitted labels
  }
}
```

Reply:
I don't fully understand all the details of this. I'll play around with it and put it in a follow-up PR.

Reply:
#6371